### PR TITLE
Implement bootstraping code for submessage support inside templates (part2).

### PR DESCRIPTION
**Pull request by guptasu**
**What this PR does / why we need it**:
**NOTE** Please read the entire description before reviewing. I can bet that it will help in the review.

This PR implements code generation of template bootstraping code (code that gets compiled in Mixer. It instantiates instances and types from config for all templates and dispatches them to adapter). Previous [part 1 PR ](https://github.com/istio/istio/pull/1617) was responsible for generating definitions of instance and type constructs from templates with sub-messages. This PR is responsible for instantiating those generated artifacts based on user config. It finishes the submessage support work.

To support submessages, the generated code is restructured in this PR so that we can do recursions. Changes are in the following generated functions : `InferType` and `ProcessCheck`, `ProcessQuota` and `ProcessReport`. The new layout is something like this : 
```
// This function creates an instance from config and dispatches to the adapter
ProcessCheck (config, adapter) {
   // Main function that build the instance. It might call into other Build functions to build sub instances for
   // the sub-messages
   BuildTemplate = func(config InstanceParam) Instance {
     var resultInst Instance
     resultInst.Value = Eval(config.Value)
     ..
     if (typeOf config.subMsgField is ResourceMessage) // this is a submessage
        instance.subMsgField = BuildSubMsg(config.subMsgField)
     }
     return resultInst
   }
   BuildSubMsg = func(config SubMsgInstanceParam) SubMsgInstance {
     var resultInst SubMsgInstance
     resultInst.field1 = Eval(config.field1)
     ..
     if (typeOf config.anotherSubMsgField is ResourceMessage) // this is a submessage
        instance.anotherSubMsgField = BuildAnotherSubMsg(config.anotherSubMsgField)
     }
     return resultInst
   }

   inst := BuildTemplate(config)

   ..
   adapter.Handle... (inst, ...)
}
```

**NOTE**: This PR is tricky to review because it is hard to visualize the diff of golang template file and the corresponding generated code. I recommend start reviewing the tests first and the baseline generated code after that (sample test files ([check.proto](https://github.com/guptasu/istio/blob/e7417cfbd3c9aa0f2b344fe1cf9224387c06cd07/mixer/tools/codegen/pkg/bootstrapgen/testdata/CheckTmpl.proto), [quota.proto](https://github.com/guptasu/istio/blob/e7417cfbd3c9aa0f2b344fe1cf9224387c06cd07/mixer/tools/codegen/pkg/bootstrapgen/testdata/QuotaTemplate.proto) and [report.proto](https://github.com/guptasu/istio/blob/e7417cfbd3c9aa0f2b344fe1cf9224387c06cd07/mixer/tools/codegen/pkg/bootstrapgen/testdata/Report1Tmpl.proto)) and baseline generated file is [AllTemplate.go.golden](https://github.com/guptasu/istio/blob/e7417cfbd3c9aa0f2b344fe1cf9224387c06cd07/mixer/tools/codegen/pkg/bootstrapgen/testdata/AllTemplates.go.golden)). 

**Also NOTE** : There are two types of tests that are implemented in this PR: 
* Baseline test under `/tools/codegen/pkg/bootstrap` which just compares the generated code with baseline view. 
* Other is run-time test for the generated code and it lives inside /template/sample/template.gen_test.go

**Release note**:
```release-note
support sub messages in Mixer template
```

### CODE REVIEW

**Comment by guptasu** [comment]
@ozevren @geeknoid @douglas-reid @mandarjog  friendly ping

**Comment by guptasu** [comment]
@geeknoid friendly ping

File: mixer/pkg/config/manager_test.go [hunk 1]
```diff
@@ -100,7 +100,7 @@ func TestConfigManager(t *testing.T) {
 func TestManager_FetchError(t *testing.T) {
 	errStr := "TestManager_FetchError"
 	store := newFakeStore("{}", "{}")
-	mgr := NewManager(nil, nil, nil, nil, nil, nil,
```
Comment (ozevren) [inline mixer/pkg/config/manager_test.go]: Nit: space?

**Comment by guptasu** [comment]
@geeknoid @ozevren re-lgtm please. I have addressed linter issues + added .gen.go files.

**Pull request merged**